### PR TITLE
Updated API

### PR DESCRIPTION
**Pull request by syrusakbary**
This PR updates the API by:
* Adds support for Rust closures
* Improving the general state of API
  * Full support for callbacks
* and covering new APIs.

Note that this is WIP. But please add any comments on things that should be improved before merge

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -14,3 +14,6 @@ edition = "2021"
 anyhow = "1.0.62"
 rusty_jsc_macros = { path = "./macros", version = "0.0.3" }
 rusty_jsc_sys = { path = "./sys", version = "0.0.3" }
+
+[dev-dependencies]
+wasmer = "3.1"
```
Comment (syrusakbary) [inline Cargo.toml:19]: This should be removed

File: examples/callback_closure.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    //     Some(value) => {
+    //         println!("{}", value.to_string(&context));
+    //     }
+    //     None => {
+    //         println!(
+    //             "Uncaught: {}",
+    //             context.get_exception().unwrap().to_string(&context)
+    //         )
+    //     }
+    // }
```
Comment (syrusakbary) [inline examples/callback_closure.rs:69]: Remove this

File: examples/properties.rs [hunk 3]
```diff
@@ -1,19 +1,180 @@
-use rusty_jsc::{JSContext, JSValue};
```
Comment (syrusakbary) [inline examples/properties.rs:1]: Rollback this file to previous version

File: src/lib.rs [hunk 4]
```diff
@@ -107,34 +122,284 @@ impl JSValue {
...
+        let num_properties = unsafe { JSPropertyNameArrayGetCount(property_name_array) };
+        let all_names = (0..num_properties)
+            .map(|property_index| {
+                JSString::from(unsafe {
+                    JSPropertyNameArrayGetNameAtIndex(property_name_array, property_index)
+                })
+                .to_string()
+            })
+            .collect::<Vec<_>>();
+        return all_names;
```
Comment (adrien-zinger) [inline src/lib.rs:376]:
```suggestion
        all_names
```

File: src/lib.rs [hunk 5]
```diff
@@ -107,34 +122,284 @@ impl JSValue {
...
         let s = JSString::from(s);
         s.to_string()
     }
+
+    // Tries to convert the value to an object
+    // TODO: This should be a Result
+    pub fn to_number(&self, context: &JSContext) -> f64 {
+        let mut exception: JSValueRef = std::ptr::null_mut();
+        let num = unsafe { JSValueToNumber(context.inner, self.inner, &mut exception) };
+        num
```
Comment (adrien-zinger) [inline src/lib.rs:138]:
```suggestion
        unsafe { JSValueToNumber(context.inner, self.inner, &mut exception) }
```

File: macros/src/lib.rs [hunk 6]
```diff
@@ -1,33 +1,83 @@
 use proc_macro::TokenStream;
 use quote::quote;
+use syn::{FnArg, Ident, Pat};
+
+fn get_name(func_argument: &FnArg) -> Ident {
+    match func_argument {
+        FnArg::Typed(fn_type) => get_name_pat(&*fn_type.pat),
+        _ => {
+            panic!("Not supported function argument: {:?}", func_argument)
```
Comment (adrien-zinger) [inline macros/src/lib.rs:9]: I think we need to add the "extra-traits" feature in syn to be able to do this.

File: src/lib.rs [hunk 7]
```diff
@@ -107,34 +122,284 @@ impl JSValue {
...
+        };
+        if !exception.is_null() {
+            return Err(JSValue::from(exception));
+        }
+        if result.is_null() {
+            panic!("Not a valid constructor");
+        }
+        // if result.is_null() {
+        //     panic!("The object has no constructor");
+        // }
```
Comment (penberg) [inline src/lib.rs:236]: drop

**Comment by syrusakbary** [comment]
I resolved most/all of the issues/concerns. I also improved the APIs to return Results when needed. Right now I think things should be shippable, since also all tests work end to end in Wasmer.

Merging!

**Pull request merged**